<a href="https://colab.research.google.com/github/dudaspm/IST210-Spring2020/blob/master/SQL/MTCars.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## MTCars

***Important***


1.   Go to File > Save a Copy to Drive (this will save to your local Google Drive)
2.   Make sure you work through this notebook from top to bottom and run each Code cell


I will take care of getting the data, starting SQL on the notebook, and creating the database. 

In [0]:
#Used to make requests
import urllib

f = urllib.request.urlopen("https://gist.githubusercontent.com/dudaspm/649cb1f241297850bbd4d27f24feb9ef/raw/31f55d399741aec488b2a0a8447405482df49168/mtcars.csv")
mtCars = str(f.read().decode('utf-8'))

In [3]:
%load_ext sql
%sql sqlite:///mtCars.db

'Connected: @mtCars.db'

***Your Turn***

Here is your ER Diagram

![weather ER Diagram](https://raw.githubusercontent.com/dudaspm/IST210-Spring2020/master/SQL/images/mtcars.png)

I've added a statement to drop (delete) the table. You need to create the table.

In [4]:
%%sql
DROP TABLE IF EXISTS Cars;
-- ## DELETE ME and add CREATE the table
CREATE Table Cars (model text primary key, mpg real, cyl real, disp real, hp real, drat real, wt real, sqec real, vs real, am real, gear real, carb real)

 * sqlite:///mtCars.db
Done.
Done.


[]

I will now insert the rest of the data (make sure your table works first by doing the step above!)

In [0]:
%%capture
for line in mtCars.split("\n")[1:]:
  d = line.split(",")
  model = d[0]
  mpg = d[1]
  cyl = d[2]
  disp = d[3]
  hp = d[4]
  drat = d[5]
  wt = d[6]
  qsec = d[7]
  vs = d[8]
  am = d[9]  
  gear = d[10]
  carb = d[11]
  %sql insert into Cars values (:model,:mpg,:cyl,:disp,:hp,:drat,:wt,:qsec,:vs,:am,:gear,:carb)

Use this select statement to make sure everything working properly. 


In [9]:
%%sql
select * from Cars limit 10;

 * sqlite:///mtCars.db
Done.


[('Mazda RX4', 21.0, 6.0, 160.0, 110.0, 3.9, 2.62, 16.46, 0.0, 1.0, 4.0, 4.0),
 ('Mazda RX4 Wag', 21.0, 6.0, 160.0, 110.0, 3.9, 2.875, 17.02, 0.0, 1.0, 4.0, 4.0),
 ('Datsun 710', 22.8, 4.0, 108.0, 93.0, 3.85, 2.32, 18.61, 1.0, 1.0, 4.0, 1.0),
 ('Hornet 4 Drive', 21.4, 6.0, 258.0, 110.0, 3.08, 3.215, 19.44, 1.0, 0.0, 3.0, 1.0),
 ('Hornet Sportabout', 18.7, 8.0, 360.0, 175.0, 3.15, 3.44, 17.02, 0.0, 0.0, 3.0, 2.0),
 ('Valiant', 18.1, 6.0, 225.0, 105.0, 2.76, 3.46, 20.22, 1.0, 0.0, 3.0, 1.0),
 ('Duster 360', 14.3, 8.0, 360.0, 245.0, 3.21, 3.57, 15.84, 0.0, 0.0, 3.0, 4.0),
 ('Merc 240D', 24.4, 4.0, 146.7, 62.0, 3.69, 3.19, 20.0, 1.0, 0.0, 4.0, 2.0),
 ('Merc 230', 22.8, 4.0, 140.8, 95.0, 3.92, 3.15, 22.9, 1.0, 0.0, 4.0, 2.0),
 ('Merc 280', 19.2, 6.0, 167.6, 123.0, 3.92, 3.44, 18.3, 1.0, 0.0, 4.0, 4.0)]

### Lab questions. 

Here are the question you will answer for the lab and you will need to:


1.   Copy the question
2.   Copy your SQL query
3.   Copy the output from the query



#### Q1

What type of cars (model) has the highest mpg for cars that 8 cylinders (cyl)?

In [23]:
%%sql
select * from Cars limit 5;

 * sqlite:///mtCars.db
Done.


[('Mazda RX4', 21.0, 6.0, 160.0, 110.0, 3.9, 2.62, 16.46, 0.0, 1.0, 4.0, 4.0),
 ('Mazda RX4 Wag', 21.0, 6.0, 160.0, 110.0, 3.9, 2.875, 17.02, 0.0, 1.0, 4.0, 4.0),
 ('Datsun 710', 22.8, 4.0, 108.0, 93.0, 3.85, 2.32, 18.61, 1.0, 1.0, 4.0, 1.0),
 ('Hornet 4 Drive', 21.4, 6.0, 258.0, 110.0, 3.08, 3.215, 19.44, 1.0, 0.0, 3.0, 1.0),
 ('Hornet Sportabout', 18.7, 8.0, 360.0, 175.0, 3.15, 3.44, 17.02, 0.0, 0.0, 3.0, 2.0)]

#### Q2

Which cars (model) have weight (wt) between 3 and 3.5, and horse power (hp) great than or equal to  123

In [24]:
%%sql
select * from Cars limit 5;

 * sqlite:///mtCars.db
Done.


[('Mazda RX4', 21.0, 6.0, 160.0, 110.0, 3.9, 2.62, 16.46, 0.0, 1.0, 4.0, 4.0),
 ('Mazda RX4 Wag', 21.0, 6.0, 160.0, 110.0, 3.9, 2.875, 17.02, 0.0, 1.0, 4.0, 4.0),
 ('Datsun 710', 22.8, 4.0, 108.0, 93.0, 3.85, 2.32, 18.61, 1.0, 1.0, 4.0, 1.0),
 ('Hornet 4 Drive', 21.4, 6.0, 258.0, 110.0, 3.08, 3.215, 19.44, 1.0, 0.0, 3.0, 1.0),
 ('Hornet Sportabout', 18.7, 8.0, 360.0, 175.0, 3.15, 3.44, 17.02, 0.0, 0.0, 3.0, 2.0)]

#### Q3

Which cars (model) have Number of forward gears (gear) 4 or 5, and mpg greater than 30?

In [25]:
%%sql
select * from Cars limit 5;

 * sqlite:///mtCars.db
Done.


[('Mazda RX4', 21.0, 6.0, 160.0, 110.0, 3.9, 2.62, 16.46, 0.0, 1.0, 4.0, 4.0),
 ('Mazda RX4 Wag', 21.0, 6.0, 160.0, 110.0, 3.9, 2.875, 17.02, 0.0, 1.0, 4.0, 4.0),
 ('Datsun 710', 22.8, 4.0, 108.0, 93.0, 3.85, 2.32, 18.61, 1.0, 1.0, 4.0, 1.0),
 ('Hornet 4 Drive', 21.4, 6.0, 258.0, 110.0, 3.08, 3.215, 19.44, 1.0, 0.0, 3.0, 1.0),
 ('Hornet Sportabout', 18.7, 8.0, 360.0, 175.0, 3.15, 3.44, 17.02, 0.0, 0.0, 3.0, 2.0)]

#### Q4

What is the difference between the highest horse power (hp) and the lowest horse power (hp)? 

In [26]:
%%sql
select * from Cars limit 5;

 * sqlite:///mtCars.db
Done.


[('Mazda RX4', 21.0, 6.0, 160.0, 110.0, 3.9, 2.62, 16.46, 0.0, 1.0, 4.0, 4.0),
 ('Mazda RX4 Wag', 21.0, 6.0, 160.0, 110.0, 3.9, 2.875, 17.02, 0.0, 1.0, 4.0, 4.0),
 ('Datsun 710', 22.8, 4.0, 108.0, 93.0, 3.85, 2.32, 18.61, 1.0, 1.0, 4.0, 1.0),
 ('Hornet 4 Drive', 21.4, 6.0, 258.0, 110.0, 3.08, 3.215, 19.44, 1.0, 0.0, 3.0, 1.0),
 ('Hornet Sportabout', 18.7, 8.0, 360.0, 175.0, 3.15, 3.44, 17.02, 0.0, 0.0, 3.0, 2.0)]

#### Q5

What is the average mpg for cars (model) with 8 cylinders (cyl) and between 100 and 200 horse power (hp)?

In [27]:
%%sql
select * from Cars limit 5;

 * sqlite:///mtCars.db
Done.


[('Mazda RX4', 21.0, 6.0, 160.0, 110.0, 3.9, 2.62, 16.46, 0.0, 1.0, 4.0, 4.0),
 ('Mazda RX4 Wag', 21.0, 6.0, 160.0, 110.0, 3.9, 2.875, 17.02, 0.0, 1.0, 4.0, 4.0),
 ('Datsun 710', 22.8, 4.0, 108.0, 93.0, 3.85, 2.32, 18.61, 1.0, 1.0, 4.0, 1.0),
 ('Hornet 4 Drive', 21.4, 6.0, 258.0, 110.0, 3.08, 3.215, 19.44, 1.0, 0.0, 3.0, 1.0),
 ('Hornet Sportabout', 18.7, 8.0, 360.0, 175.0, 3.15, 3.44, 17.02, 0.0, 0.0, 3.0, 2.0)]

#### Q6

What is the average mpg and the number of cars (model) that are automatic (am is 0)

In [28]:
%%sql
select * from Cars limit 5;

 * sqlite:///mtCars.db
Done.


[('Mazda RX4', 21.0, 6.0, 160.0, 110.0, 3.9, 2.62, 16.46, 0.0, 1.0, 4.0, 4.0),
 ('Mazda RX4 Wag', 21.0, 6.0, 160.0, 110.0, 3.9, 2.875, 17.02, 0.0, 1.0, 4.0, 4.0),
 ('Datsun 710', 22.8, 4.0, 108.0, 93.0, 3.85, 2.32, 18.61, 1.0, 1.0, 4.0, 1.0),
 ('Hornet 4 Drive', 21.4, 6.0, 258.0, 110.0, 3.08, 3.215, 19.44, 1.0, 0.0, 3.0, 1.0),
 ('Hornet Sportabout', 18.7, 8.0, 360.0, 175.0, 3.15, 3.44, 17.02, 0.0, 0.0, 3.0, 2.0)]